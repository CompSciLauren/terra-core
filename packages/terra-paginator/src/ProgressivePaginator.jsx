import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import ResponsiveElement from 'terra-responsive-element';
import KeyCode from 'keycode-js';
import IconPrevious from 'terra-icon/lib/icon/IconPrevious';
import IconNext from 'terra-icon/lib/icon/IconNext';
import Button from '../../terra-button';
import styles from './Paginator.module.scss';

import { calculatePages } from './_paginationUtils';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Function to be executed when a navigation element is selected.
   */
  onPageChange: PropTypes.func.isRequired,
  /**
   * The active/selected page.
   */
  selectedPage: PropTypes.number.isRequired,
  /**
   * Total number of all items being paginated.
   */
  totalCount: PropTypes.number.isRequired,
  /**
   * Total number of items per page.
   */
  itemCountPerPage: PropTypes.number.isRequired,
};

const contextTypes = {
  /* eslint-disable consistent-return */
  intl: (context) => {
    if (context.intl === undefined) {
      return new Error('Component is internationalized, and must be wrapped in terra-base');
    }
  },
};

class ProgressivePaginator extends React.Component {
  constructor(props) {
    super(props);

    const { selectedPage, totalCount } = this.props;

    this.state = {
      selectedPage: selectedPage && totalCount ? selectedPage : undefined,
    };

    this.handlePageChange = this.handlePageChange.bind(this);
    this.handleOnKeyDown = this.handleOnKeyDown.bind(this);
    this.defaultProgressivePaginator = this.defaultProgressivePaginator.bind(this);
    this.reducedProgressivePaginator = this.reducedProgressivePaginator.bind(this);
  }

  handlePageChange(index) {
    return (event) => {
      event.preventDefault();

      this.props.onPageChange(index);
      this.setState({ selectedPage: index });
    };
  }

  handleOnKeyDown(index) {
    return (event) => {
      if (event.nativeEvent.keyCode === KeyCode.KEY_RETURN || event.nativeEvent.keyCode === KeyCode.KEY_SPACE) {
        event.preventDefault();

        if (Number.isNaN(Number(index))) {
          this.props.onPageChange(event.target.text.trim().toLowerCase());

          return false;
        }

        this.props.onPageChange(index);
        this.setState({
          selectedPage: index,
        });
      }

      return false;
    };
  }

  defaultProgressivePaginator() {
    const totalPages = calculatePages(this.props.totalCount, this.props.itemCountPerPage);
    const { selectedPage } = this.state;
    const previousPageIndex = selectedPage === 1 ? 1 : selectedPage - 1;
    const nextPageIndex = selectedPage === totalPages ? totalPages : selectedPage + 1;
    const areLeftControlsDisabled = selectedPage === 1;
    const areRightControlsDisabled = selectedPage === totalPages;

    return (
      <div className={cx(['paginator', 'progressive'])} role="navigation" aria-label="pagination">
        <div>
          {this.context.intl.formatMessage({ id: 'Terra.paginator.pageCount' }, { selectedPage, totalPages })}
        </div>
        <div>
          <Button
            className={cx(['nav-link', areLeftControlsDisabled && 'is-disabled'])}
            isDisabled={areLeftControlsDisabled}
            aria-disabled={areLeftControlsDisabled}
            aria-label="first"
            tabIndex={areLeftControlsDisabled ? null : '0'}
            onClick={this.handlePageChange(1)}
            onKeyDown={this.handleOnKeyDown(1)}
            text={this.context.intl.formatMessage({ id: 'Terra.paginator.first' })}
          />
          <Button
            className={cx(['nav-link', 'previous', areLeftControlsDisabled && 'is-disabled'])}
            icon={<IconPrevious />}
            isDisabled={areLeftControlsDisabled}
            aria-disabled={areLeftControlsDisabled}
            aria-label="previous"
            tabIndex={areLeftControlsDisabled ? null : '0'}
            onClick={this.handlePageChange(previousPageIndex)}
            onKeyDown={this.handleOnKeyDown(previousPageIndex)}
            text={this.context.intl.formatMessage({ id: 'Terra.paginator.previous' })}
          />
          <Button
            className={cx(['nav-link', 'next', areRightControlsDisabled && 'is-disabled'])}
            icon={<IconNext />}
            isReversed
            isDisabled={areRightControlsDisabled}
            aria-disabled={areRightControlsDisabled}
            aria-label="next"
            tabIndex={areRightControlsDisabled ? null : '0'}
            onClick={this.handlePageChange(nextPageIndex)}
            onKeyDown={this.handleOnKeyDown(nextPageIndex)}
            text={this.context.intl.formatMessage({ id: 'Terra.paginator.next' })}
          />
          <Button
            className={cx(['nav-link', areRightControlsDisabled && 'is-disabled'])}
            isDisabled={areRightControlsDisabled}
            aria-disabled={areRightControlsDisabled}
            aria-label="last"
            tabIndex={areRightControlsDisabled ? null : '0'}
            onClick={this.handlePageChange(totalPages)}
            onKeyDown={this.handleOnKeyDown(totalPages)}
            text={this.context.intl.formatMessage({ id: 'Terra.paginator.last' })}
          />
        </div>
      </div>
    );
  }

  reducedProgressivePaginator() {
    const totalPages = calculatePages(this.props.totalCount, this.props.itemCountPerPage);
    const { selectedPage } = this.state;
    const previousPageIndex = selectedPage === 1 ? 1 : selectedPage - 1;
    const nextPageIndex = selectedPage === totalPages ? totalPages : selectedPage + 1;
    const areLeftControlsDisabled = selectedPage === 1;
    const areRightControlsDisabled = selectedPage === totalPages;

    return (
      <div className={cx(['paginator'])} role="navigation" aria-label="pagination">
        <div>
          <Button
            className={cx(['nav-link', areLeftControlsDisabled && 'is-disabled'])}
            isDisabled={areLeftControlsDisabled}
            aria-disabled={areLeftControlsDisabled}
            aria-label="first"
            tabIndex={areLeftControlsDisabled ? null : '0'}
            onClick={this.handlePageChange(1)}
            onKeyDown={this.handleOnKeyDown(1)}
            text={this.context.intl.formatMessage({ id: 'Terra.paginator.first' })}
          />
          <Button
            className={cx(['nav-link', 'previous', areLeftControlsDisabled && 'is-disabled'])}
            icon={<IconPrevious />}
            isIconOnly
            isDisabled={areLeftControlsDisabled}
            aria-disabled={areLeftControlsDisabled}
            aria-label="previous"
            tabIndex={areLeftControlsDisabled ? null : '0'}
            onClick={this.handlePageChange(previousPageIndex)}
            onKeyDown={this.handleOnKeyDown(previousPageIndex)}
            text={this.context.intl.formatMessage({ id: 'Terra.paginator.previous' })}
          />
        </div>
        <div>
          {this.context.intl.formatMessage({ id: 'Terra.paginator.pageCount' }, { selectedPage, totalPages })}
        </div>
        <div>
          <Button
            className={cx(['nav-link', 'next', areRightControlsDisabled && 'is-disabled'])}
            icon={<IconNext />}
            isIconOnly
            isDisabled={areRightControlsDisabled}
            aria-disabled={areRightControlsDisabled}
            aria-label="next"
            tabIndex={areRightControlsDisabled ? null : '0'}
            onClick={this.handlePageChange(nextPageIndex)}
            onKeyDown={this.handleOnKeyDown(nextPageIndex)}
            text={this.context.intl.formatMessage({ id: 'Terra.paginator.next' })}
          />
          <Button
            className={cx(['nav-link', areRightControlsDisabled && 'is-disabled'])}
            isDisabled={areRightControlsDisabled}
            aria-disabled={areRightControlsDisabled}
            aria-label="last"
            tabIndex={areRightControlsDisabled ? null : '0'}
            onClick={this.handlePageChange(totalPages)}
            onKeyDown={this.handleOnKeyDown(totalPages)}
            text={this.context.intl.formatMessage({ id: 'Terra.paginator.last' })}
          />
        </div>
      </div>
    );
  }

  render() {
    return <ResponsiveElement defaultElement={this.reducedProgressivePaginator()} tiny={this.defaultProgressivePaginator()} />;
  }
}

ProgressivePaginator.propTypes = propTypes;
ProgressivePaginator.contextTypes = contextTypes;

export default ProgressivePaginator;
