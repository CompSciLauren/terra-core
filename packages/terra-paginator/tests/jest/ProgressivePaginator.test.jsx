import React from 'react';
/* eslint-disable-next-line import/no-extraneous-dependencies */
import { mountWithIntl } from 'terra-enzyme-intl';

import ProgressivePaginator from '../../src/ProgressivePaginator';

describe('ProgressivePaginator', () => {
  // Snapshot Tests
  it('should render a default ProgressivePaginator', () => {
    const wrapper = mountWithIntl(<ProgressivePaginator onPageChange={e => typeof e} selectedPage={1} totalCount={2234} itemCountPerPage={20} />);
    expect(wrapper).toMatchSnapshot();
  });
});
