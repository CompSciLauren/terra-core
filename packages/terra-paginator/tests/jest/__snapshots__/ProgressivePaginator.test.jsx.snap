// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressivePaginator should render a default ProgressivePaginator 1`] = `
<ProgressivePaginator
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": null,
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  itemCountPerPage={20}
  onPageChange={[Function]}
  selectedPage={1}
  totalCount={2234}
>
  <ResponsiveElement
    defaultElement={
      <div
        aria-label="pagination"
        className="paginator"
        role="navigation"
      >
        <div>
          <button
            aria-disabled={true}
            aria-label="first"
            className="nav-link is-disabled"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={null}
            type="button"
          >
            Terra.paginator.first
          </button>
          <button
            aria-disabled={true}
            aria-label="previous"
            className="nav-link previous icon-only is-disabled"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={null}
            type="button"
          >
            <span
              className="visually-hidden"
            >
              Previous
            </span>
            <span
              className="icon"
            />
          </button>
        </div>
        <div>
          Terra.paginator.pageCount
        </div>
        <div>
          <button
            aria-disabled={false}
            aria-label="next"
            className="nav-link next icon-only"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="visually-hidden"
            >
              Next
            </span>
            <span
              className="icon"
            />
          </button>
          <button
            aria-disabled={false}
            aria-label="last"
            className="nav-link"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex="0"
            type="button"
          >
            Terra.paginator.last
          </button>
        </div>
      </div>
    }
    responsiveTo="parent"
    tiny={
      <div
        aria-label="pagination"
        className="paginator progressive"
        role="navigation"
      >
        <div>
          Terra.paginator.pageCount
        </div>
        <div>
          <button
            aria-disabled={true}
            aria-label="first"
            className="nav-link is-disabled"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={null}
            type="button"
          >
            Terra.paginator.first
          </button>
          <button
            aria-disabled={true}
            aria-label="previous"
            className="nav-link previous is-disabled"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={null}
            type="button"
          >
            <span
              className="icon"
            />
            Terra.paginator.previous
          </button>
          <button
            aria-disabled={false}
            aria-label="next"
            className="nav-link next"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex="0"
            type="button"
          >
            Terra.paginator.next
            <span
              className="icon"
            />
          </button>
          <button
            aria-disabled={false}
            aria-label="last"
            className="nav-link"
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex="0"
            type="button"
          >
            Terra.paginator.last
          </button>
        </div>
      </div>
    }
  >
    <div />
  </ResponsiveElement>
</ProgressivePaginator>
`;
